ARCHITECTURE.md

```diff
@@ -180,16 +180,80 @@ Issues with this approach:
 
 * Some stuff from Core won’t be used in React (or won’t be needed by all users), `updateAll()` for example, that is tied to current UI rendering implementation. *Might not be an issue if it’s just a few methods.*
 
-C. Another approach is to create a separate version of Uppy for React, uppy-react, not use Core in it, re-create Core in form of UppyContainer, manage state in that (and possibly allow for Redux, Mobx or whatever else someone is using), while re-using some parts of current Uppy. This has been explored here: https://github.com/hedgerh/uppy-react.
-
-Ideally we would like to not end up with two separate versions — regular Uppy and React Uppy — that needs to be maintained and both updated each time. Though sometimes that’s exactly what lib authors are doing — creating separate React versions.
-
----
-
 Also, Yo-Yo (Bel) has some issues:
 
 - Local network requests from `<img src="">` are made on each state update. Not a big deal, but annoying and does not happing with vdom solutions, like Preact.
 - Webcam currently flashes when state is updated (morphdom issue?).
 - When using template strings it might be harder to re-use UI components in React, where JSX is a standard.
 
 We are thinking about trying Preact for more stability and React compatibility, but not sure if it’s worth it.
+
+
+## Harry's Proposal
+---
+My proposal for Uppy's architecture and API address two use cases separately:
+
+1. Uppy when its used without React and state is managed internally
+
+2. Uppy when it's used with React, and the user manages their own state (or we provide a React container component that takes care of it for them.)  I'll refer to them as the "vanilla/yo-yo" and "React" use cases, respectively.
+
+### Vanilla (using yo-yo) with managed state
+
+So we have a problem. When we decouple rendering from Uppy, and a user is able to make their own custom UI, they'll need to be able to call on plugin methods to do things, like fetch a list of files from Google Drive, for example.  In our current setup, the core creates and holds an instance of every plugin, and outside UIs don't have direct access to these plugin instances.  To get around this, one idea we was to expose a bus/event emitter, and they could emit something like `bus.emit('google-drive:list-files', { directory: 'root' })` when they wanted to fetch a list of files.
+
```

Comment:
Yes, this is an issue. Plus you can’t get a response after emitting event, but same with dispatcher. So it only updates state, which I guess is fine and aligns with data down — actions up approach. Instead of long actions like this we could allow accessing methods through core or actions shortcut, [like in FineUploader-React here](https://github.com/FineUploader/react-fine-uploader/blob/master/src/components/delete-button.jsx#L44). Like this: `actions.googleDrive.listFiles({ directory: 'root' })`.